pinot-core/src/main/java/org/apache/pinot/core/operator/query/DictionaryBasedDistinctOperator.java

```diff
@@ -0,0 +1,146 @@
+      Dictionary dictionary, int numTotalDocs,
+      TransformOperator transformOperator) {
+    super(indexSegment, distinctAggregationFunction, transformOperator);
+
+    _distinctAggregationFunction = distinctAggregationFunction;
+    _dictionary = dictionary;
+    _numTotalDocs = numTotalDocs;
+
+    List<OrderByExpressionContext> orderByExpressionContexts = _distinctAggregationFunction.getOrderByExpressions();
+
+    if (orderByExpressionContexts != null) {
+      OrderByExpressionContext orderByExpressionContext = orderByExpressionContexts.get(0);
+
+      _isAscending = orderByExpressionContext.isAsc();
+      _hasOrderBy = true;
+    }
+
+    _dictLength = _dictionary.length();
+  }
+
+  @Override
+  protected IntermediateResultsBlock getNextBlock() {
+    DistinctTable distinctTable = buildResult();
+
+    return new IntermediateResultsBlock(new AggregationFunction[]{_distinctAggregationFunction},
+        Collections.singletonList(distinctTable), false);
+  }
+
+  /**
+   * Build the final result for this operation
+   */
+  private DistinctTable buildResult() {
+
+    assert _distinctAggregationFunction.getType() == AggregationFunctionType.DISTINCT;
+
+    List<ExpressionContext> expressions = _distinctAggregationFunction.getInputExpressions();
+    ExpressionContext expression = expressions.get(0);
+    FieldSpec.DataType dataType = _dictionary.getValueType();
+
+    DataSchema dataSchema = new DataSchema(new String[]{expression.toString()},
+        new DataSchema.ColumnDataType[]{DataSchema.ColumnDataType.fromDataTypeSV(dataType)});
+    List<Record> records;
+
+    int limit = _distinctAggregationFunction.getLimit();
+    int actualLimit = Math.min(limit, _dictLength);
+
+    // If ORDER BY is not present, we read the first limit values from the dictionary and return.
+    // If ORDER BY is present and the dictionary is sorted, then we read the first/last limit values
+    // from the dictionary. If not sorted, then we read the entire dictionary and return it.
+    if (!_hasOrderBy) {
+      records = new ArrayList<>(actualLimit);
+
+      for (int i = 0; i < actualLimit; i++) {
+        records.add(new Record(new Object[]{_dictionary.getInternal(i)}));
+      }
+    } else {
+      DistinctTable distinctTable = new DistinctTable(dataSchema, _distinctAggregationFunction.getOrderByExpressions(), limit);
+
+      if (_dictionary.isSorted()) {
+        if (_isAscending) {
+          for (int i = 0; i < actualLimit; i++) {
+            distinctTable.addWithOrderBy(new Record(new Object[]{_dictionary.getInternal(i)}));
+          }
+        } else {
+          for (int i = _dictLength - 1; i >= (_dictLength - actualLimit); i--) {
+            distinctTable.addWithOrderBy(new Record(new Object[]{_dictionary.getInternal(i)}));
+          }
+        }
+      } else {
+        for (int i = 0; i < _dictLength; i++) {
```

Comment:
You only need to create main distinct table when dictionary is not sorted. When dictionary is sorted, creating a wrapper distinct table is good enough